File: mockAnnotations/src/main/java/com/azazellj/mock/annotations/internal/MockUtils.java
package com.azazellj.mock.annotations.internal;

import com.azazellj.mock.annotations.MockTypeHelper;

import java.util.Collection;
import java.util.Map;
import java.util.UUID;

public class MockUtils {

    @MockTypeHelper(types = {boolean.class, Boolean.class})
    public static boolean randomBoolean() {
        return Math.random() < 0.5;
    }

    @MockTypeHelper(types = {byte.class, Byte.class})
    public static byte randomByte() {
        return (byte) (Math.random() * minOrMax(Byte.MAX_VALUE, Byte.MIN_VALUE).byteValue());
    }

    @MockTypeHelper(types = {char.class, Character.class})
    public static char randomChar() {
        return randomString().charAt((int) (Math.random() * 16));
    }

    @MockTypeHelper(types = {double.class, Double.class})
    public static double randomDouble() {
        return (Math.random() * minOrMax((long) Double.MAX_VALUE, (long) Double.MIN_VALUE).doubleValue());
    }

    @MockTypeHelper(types = {float.class, Float.class})
    public static float randomFloat() {
        return (float) (Math.random() * minOrMax((long) Float.MAX_VALUE, (long) Float.MIN_VALUE).floatValue());
    }

    @MockTypeHelper(types = {int.class, Integer.class})
    public static int randomInt() {
        return (int) (Math.random() * minOrMax(Integer.MAX_VALUE, Integer.MIN_VALUE).intValue());
    }

    @MockTypeHelper(types = {long.class, Long.class})
    public static long randomLong() {
        return (long) (Math.random() * minOrMax(Long.MAX_VALUE, Long.MIN_VALUE).longValue());
    }

    @MockTypeHelper(types = {short.class, Short.class})
    public static short randomShort() {
        return (short) (Math.random() * minOrMax(Short.MAX_VALUE, Short.MIN_VALUE).shortValue());
    }

    private static Number minOrMax(long max, long min) {
        return randomBoolean() ? max : min;
    }

    @MockTypeHelper(types = String.class)
    public static String randomString() {
        return UUID.randomUUID().toString();
    }

    @MockTypeHelper(types = Collection.class, initNeeded = true)
    public static <T, C extends Collection<T>> C randomCollection(C collection, T randomValue) {
        collection.add(randomValue);
        return collection;
    }

    @MockTypeHelper(types = Map.class, initNeeded = true)
    public static <K, V, M extends Map<K, V>> M randomMap(M map, K key, V value) {
        map.put(key, value);
        return map;
    }
}

File: mockAnnotations-kt/src/main/kotlin/com/azazellj/mock/annotations/internal/MockUtils.kt
package com.azazellj.mock.annotations.internal

import com.azazellj.mock.annotations.MockTypeHelper
import java.util.*

object MockUtils {
    @MockTypeHelper(types = arrayOf(Boolean::class))
    fun randomBoolean(): Boolean {
        return Math.random() < 0.5
    }

    @MockTypeHelper(types = [(Byte::class)])
    fun randomByte(): Byte {
        return (Math.random() * minOrMax(java.lang.Byte.MAX_VALUE.toLong(), java.lang.Byte.MIN_VALUE.toLong()).toByte()).toByte()
    }

    @MockTypeHelper(types = [(Char::class)])
    fun randomChar(): Char {
        return randomString()[(Math.random() * 16).toInt()]
    }

    @MockTypeHelper(types = [(Double::class)])
    fun randomDouble(): Double {
        return Math.random() * minOrMax(java.lang.Double.MAX_VALUE.toLong(), java.lang.Double.MIN_VALUE.toLong()).toDouble()
    }

    @MockTypeHelper(types = [(Float::class)])
    fun randomFloat(): Float {
        return (Math.random() * minOrMax(java.lang.Float.MAX_VALUE.toLong(), java.lang.Float.MIN_VALUE.toLong()).toFloat()).toFloat()
    }

    @MockTypeHelper(types = [(Int::class)])
    fun randomInt(): Int {
        return (Math.random() * minOrMax(Integer.MAX_VALUE.toLong(), Integer.MIN_VALUE.toLong()).toInt()).toInt()
    }

    @MockTypeHelper(types = [(Long::class)])
    fun randomLong(): Long {
        return (Math.random() * minOrMax(java.lang.Long.MAX_VALUE, java.lang.Long.MIN_VALUE).toLong()).toLong()
    }

    @MockTypeHelper(types = [(Short::class)])
    fun randomShort(): Short {
        return (Math.random() * minOrMax(java.lang.Short.MAX_VALUE.toLong(), java.lang.Short.MIN_VALUE.toLong()).toShort()).toShort()
    }

    private fun minOrMax(max: Long, min: Long): Number {
        return if (randomBoolean()) max else min
    }

    @MockTypeHelper(types = [(String::class)])
    fun randomString(): String {
        return UUID.randomUUID().toString()
    }

    @MockTypeHelper(types = [MutableCollection::class], initNeeded = true)
    fun <T, C : MutableCollection<T>> randomCollection(collection: C, randomValue: T): C {
        collection.add(randomValue)
        return collection
    }

    @MockTypeHelper(types = [MutableMap::class], initNeeded = true)
    fun <K, V, M : MutableMap<K, V>> randomMap(map: M, key: K, value: V): M {
        map.put(key, value)
        return map
    }
}